File: src/dotnet/Harness.Engine/Evaluators.cs
using System.Text.RegularExpressions;

namespace Harness.Engine;

/// <summary>
/// Deterministic evaluators (Exact Match, Regex, Trajectory) — the part of the diagram's
/// Evaluator #6 that does NOT need an LLM. They run in-process over <see cref="Trace"/>
/// and <see cref="HarnessState"/>, cost zero tokens, and serve as a gate: only when they
/// pass is it worth escalating to the LLM judge (a saving under the token budget).
/// </summary>
public static class Evaluators
{
    public static Score ExactMatch(string expected, string actual) =>
        new("exact_match", Norm(expected) == Norm(actual) ? 1.0 : 0.0,
            $"expected=\"{expected}\" actual=\"{actual}\"");

    public static Score MatchesRegex(string pattern, string actual) =>
        new("regex", Regex.IsMatch(actual, pattern) ? 1.0 : 0.0, pattern);

    /// <summary>
    /// Fraction of the expected prefix that matched, in order. An out-of-sequence step
    /// cuts the count right there — trajectory is about path, not set.
    /// </summary>
    public static Score Trajectory(IReadOnlyList<string> expected, IReadOnlyList<string> actual)
    {
        var matched = 0;
        for (var i = 0; i < expected.Count && i < actual.Count; i++)
        {
            if (expected[i] != actual[i])
                break;
            matched++;
        }

        var value = expected.Count == 0 ? 1.0 : (double)matched / expected.Count;
        return new("trajectory", value, $"{matched}/{expected.Count} steps in the expected order");
    }

    /// <summary>Were all expected domain keys filled in the final state?</summary>
    public static Score Completeness(HarnessState state, IReadOnlyList<string> requiredKeys)
    {
        var filled = requiredKeys.Count(k => !string.IsNullOrWhiteSpace(state.Data.GetValueOrDefault(k, "")));
        var value = requiredKeys.Count == 0 ? 1.0 : (double)filled / requiredKeys.Count;
        return new("completeness", value, $"{filled}/{requiredKeys.Count} keys filled");
    }

    /// <summary>
    /// Ended in <see cref="TraceOutcome.Stop"/> without hitting the step ceiling or the
    /// time one (<see cref="TraceOutcome.Timeout"/>) — both would be indistinguishable
    /// from a simply incomplete trajectory if not checked separately.
    /// </summary>
    public static Score StepBudget(IReadOnlyList<TraceEntry> trace)
    {
        var hitBudget = trace.Any(e => e.Outcome == TraceOutcome.Budget);
        var hitTimeout = trace.Any(e => e.Outcome == TraceOutcome.Timeout);
        var terminated = trace.Any(e => e.Outcome == TraceOutcome.Stop);

        return new("budget", !hitBudget && !hitTimeout && terminated ? 1.0 : 0.0,
            hitBudget ? "cut off by the step ceiling"
            : hitTimeout ? "cut off by the time ceiling (timeout)"
            : terminated ? "completed within budget"
            : "did not finish");
    }

    /// <summary>Trace commands in order, skipping corrective-error turns by default.</summary>
    public static IReadOnlyList<string> CommandsOf(IReadOnlyList<TraceEntry> trace, bool includeErrors = false) =>
        trace.Where(e => includeErrors || e.Outcome != TraceOutcome.Error)
             .Select(e => e.Command)
             .ToList();

    private static string Norm(string value) => value.Trim();
}

/// <summary>Score of a metric in [0,1]. <see cref="Passed"/> requires a full match.</summary>
public record Score(string Metric, double Value, string Detail = "")
{
    public bool Passed => Value >= 1.0;
}

File: src/go/harnessengine/evaluators.go
package harnessengine

import (
	"fmt"
	"regexp"
	"strings"
)

// Deterministic evaluators (Exact Match, Regex, Trajectory) — the diagram's Evaluator that
// does NOT need an LLM. They run in-process over Trace and HarnessState, cost zero tokens,
// and serve as a gate: only when they pass is it worth escalating to the LLM judge (a
// saving under the token budget).

// Score is the grade for a metric in [0,1]. Passed requires a full match.
type Score struct {
	Metric string
	Value  float64
	Detail string
}

// Passed reports whether the score reached the maximum (1.0).
func (s Score) Passed() bool {
	return s.Value >= 1.0
}

// ExactMatch compares expected and actual, trimmed.
func ExactMatch(expected, actual string) Score {
	value := 0.0
	if strings.TrimSpace(expected) == strings.TrimSpace(actual) {
		value = 1.0
	}
	return Score{Metric: "exact_match", Value: value, Detail: fmt.Sprintf(`esperado="%s" obtido="%s"`, expected, actual)}
}

// MatchesRegex checks whether actual matches pattern.
func MatchesRegex(pattern, actual string) Score {
	value := 0.0
	if matched, err := regexp.MatchString(pattern, actual); err == nil && matched {
		value = 1.0
	}
	return Score{Metric: "regex", Value: value, Detail: pattern}
}

// Trajectory scores the fraction of the expected prefix that matched, in order. A
// step out of sequence stops the count there — trajectory is about the path, not the set.
func Trajectory(expected, actual []string) Score {
	matched := 0
	for i := 0; i < len(expected) && i < len(actual); i++ {
		if expected[i] != actual[i] {
			break
		}
		matched++
	}

	value := 1.0
	if len(expected) > 0 {
		value = float64(matched) / float64(len(expected))
	}
	return Score{Metric: "trajectory", Value: value, Detail: fmt.Sprintf("%d/%d passos na ordem esperada", matched, len(expected))}
}

// Completeness scores the fraction of required domain keys that were filled in the final state.
func Completeness(state HarnessState, requiredKeys []string) Score {
	filled := 0
	for _, k := range requiredKeys {
		if strings.TrimSpace(state.Data[k]) != "" {
			filled++
		}
	}
	value := 1.0
	if len(requiredKeys) > 0 {
		value = float64(filled) / float64(len(requiredKeys))
	}
	return Score{Metric: "completeness", Value: value, Detail: fmt.Sprintf("%d/%d chaves preenchidas", filled, len(requiredKeys))}
}

// StepBudget reports whether the run ended in TraceOutcome.Stop without hitting the step
// ceiling nor the time ceiling (TraceOutcome.Timeout) — both would be indistinguishable
// from a simply-incomplete trajectory if not checked separately.
func StepBudget(trace []TraceEntry) Score {
	hitBudget, hitTimeout, terminated := false, false, false
	for _, e := range trace {
		switch e.Outcome {
		case TraceOutcome.Budget:
			hitBudget = true
		case TraceOutcome.Timeout:
			hitTimeout = true
		case TraceOutcome.Stop:
			terminated = true
		}
	}

	value := 0.0
	detail := "não terminou"
	switch {
	case hitBudget:
		detail = "cortado pelo teto de passos"
	case hitTimeout:
		detail = "cortado pelo teto de tempo (timeout)"
	case terminated:
		detail = "concluído dentro do teto"
	}
	if !hitBudget && !hitTimeout && terminated {
		value = 1.0
	}

	return Score{Metric: "budget", Value: value, Detail: detail}
}

// CommandsOf returns the trace's commands in order, by default skipping corrective-error turns.
func CommandsOf(trace []TraceEntry, includeErrors bool) []string {
	commands := make([]string, 0, len(trace))
	for _, e := range trace {
		if includeErrors || e.Outcome != TraceOutcome.Error {
			commands = append(commands, e.Command)
		}
	}
	return commands
}
